### PR TITLE
Add feature parsing form data

### LINKED ISSUE #1: Proposing to make improvements to the curl parsing
The current implementation of the Curl class does not correctly handle form data (-F flag) in cURL commands. Specifically, multipart form data is not being parsed or included in the resulting Curl object, which limits support for requests requiring file uploads or complex form submissions. This issue affects the ability to fully represent cURL commands in Dart

I was thinking of improving the parsing by allowing multiple -F flags and then returning an additional formFields Map<String,dynamic> but this will have an issue. As you might know sending out lists in formdata we are required to do it like 

image

image

So if we like to parse the curl it will be having curl .... -F "arr=ele1" -F "arr=ele2"

which should result in a map of arr = [ele1,ele2]

Hence we wouldn't be able to return formData as a Map because of the non-unique key therefore I think a list of tuples will be better or a list of formdata class objects

This can be achieved easily I am just writing up my approach.

Suggestions?

### LINKED ISSUE #503: [curl_parser] Improvement to the way curl parsing handles form-data
The current implementation of the Curl class does not correctly handle form data (-F flag) in cURL commands. Specifically, multipart form data is not being parsed or included in the resulting Curl object, which limits support for requests requiring file uploads or complex form submissions. This issue affects the ability to fully represent cURL commands in Dart

I was thinking of improving the parsing by allowing multiple -F flags and then returning an additional formFields Map<String,dynamic> but this will have an issue. As you might know sending out lists in formdata we are required to do it like 

image

image

So if we like to parse the curl it will be having curl .... -F "arr=ele1" -F "arr=ele2"

which should result in a map of arr = [ele1,ele2]

Hence we wouldn't be able to return formData as a Map because of the non-unique key therefore I think a list of tuples will be better or a list of formdata class objects

This can be achieved easily I am just writing up my approach.

Suggestions?

### PR DESCRIPTION
**Pull request by Pratham-commits-code**
# Add Form Data Parsing Support for cURL Commands

## Overview
This PR adds support for parsing form data sent using the `-F` flag in cURL commands. The implementation handles both regular form fields and file uploads while maintaining package compatibility across different Dart versions.

### Data Structure Choice
For storing form data, I implemented a `List<List<String>>` structure where each inner list represents a key-value pair. The alternatives considered were:

1. **FormData class**: Rejected because it would introduce a web dependency through `dart:html` or the new `web` package
2. **Records (key-value pairs)**: Rejected as it would require Dart >=3.0, limiting compatibility
3. **Map<String, String>**: Rejected as it would negate the ability to add lists.

The chosen implementation:
- Maintains package compatibility
- Preserves field order
- Uses an assert statement to ensure each inner list has exactly 2 elements (key and value)

### File Upload Handling
- Detects file uploads by checking for the `@` prefix in the value
- Strips the `@` prefix from file paths
- Preserves the original file path in the value field

## Example Usage

```dart
// Parsing a cURL command with form data
const curl = '''
curl -X POST https://example.com/upload \\
  -F "file=@/path/to/image.jpg" \\
  -F "username=john"
''';

final parsed = Curl.parse(curl);
// Returns:
// formData: [
//   ['file', '/path/to/image.jpg'],
//   ['username', 'john']
// ]
```

## Testing
Added test cases covering:
- Basic form field parsing
- File upload parsing
- Multiple form fields
- Mixed regular fields and file uploads
- Edge cases with invalid form data

## Breaking Changes
None. This is a backward-compatible addition to the existing functionality.

## Future Considerations
1. Consider adding type safety for form fields in future versions
2. Potential support for more complex form data structures when Dart 3.0 becomes the minimum supported version
3. Possible addition of mime-type detection for file uploads

## Related Issues
Closes foss42/apidash#503 

## Checklist
- [x] Added implementation for form data parsing
- [x] Added comprehensive tests
- [x] Updated documentation
- [x] Maintained backward compatibility
- [x] Added appropriate assertions for data validation

### CODE REVIEW

**Comment by Pratham-commits-code** [comment]
I was thinking of not removing the '@' from the @/file/path. Thoughts?

File: lib/src/models/curl.dart [hunk 1]
```diff
@@ -128,40 +150,51 @@ class Curl extends Equatable {
...
-    String? url = clean(result['url']);
-    final String? data = result['data'];
-    final String? cookie = result['cookie'];
-    final String? user = result['user'];
-    final String? referer = result['referer'];
-    final String? userAgent = result['user-agent'];
-    final bool form = result['form'] ?? false;
-    final bool head = result['head'] ?? false;
-    final bool insecure = result['insecure'] ?? false;
-    final bool location = result['location'] ?? false;
```
Comment (ashitaprasad) [inline lib/src/models/curl.dart:140]: Revert and use this approach of accessing values and storing in local variables.

File: lib/src/models/curl.dart [hunk 2]
```diff
@@ -128,40 +150,51 @@ class Curl extends Equatable {
...
-      location: location,
+      data: result['data'],
+      cookie: result['cookie'],
+      user: result['user'],
+      referer: result['referer'],
+      userAgent: result['user-agent'],
+      form: formData != null && formData.isNotEmpty,
+      formData: formData,
+      insecure: result['insecure'] ?? false,
+      location: result['location'] ?? false,
```
Comment (ashitaprasad) [inline lib/src/models/curl.dart]: Accessing dict values during object creation is not preferred. 
Revert to supplying values via local variables as it was done earlier.

File: lib/src/models/curl.dart [hunk 3]
```diff
@@ -34,6 +36,10 @@ class Curl extends Equatable {
   /// Sends data as a multipart/form-data request.
   final bool form;
 
+  /// Form data list.
+  /// Currently, it is represented as a list of key-value pairs ([key, value]).
+  final List<List<String>>? formData;
```
Comment (ashitaprasad) [inline lib/src/models/curl.dart]: It should be a `List<FormDataModel>`
See [here](https://github.com/foss42/apidash/blob/main/packages/apidash_core/lib/models/form_data_model.dart)
It will take care of files as well.

File: test/curl_converter_test.dart [hunk 4]
```diff
@@ -27,6 +27,97 @@ void main() {
     );
   }, timeout: defaultTimeout);
```
Comment (ashitaprasad) [inline test/curl_converter_test.dart:30]: Add more tests for Curl.parse 
which handle fields with array values, headers, etc.

**Comment by ashitaprasad** [comment]
> I was thinking of not removing the '@' from the @/file/path. Thoughts?

Follow the review comments for a better solution.

File: lib/src/models/curl.dart (hunk 3 above)
Comment (Pratham-commits-code) [reply lib/src/models/curl.dart]: I thought about doing that but didn't because U thought it will just add a dependency. Will surely do that now.

Comment (ashitaprasad) [reply lib/src/models/curl.dart]: No dependency must be added. You can create a copy of the model file ... It will be refactored later.

**Comment by ashitaprasad** [comment]
The package has been moved to https://github.com/foss42/apidash/tree/main/packages/curl_parser
I have transferred the issue to the repo. 
You can open up a new PR with the mentioned changes there.
